kevm_pyk/src/kevm_pyk/solc_to_k.py

```diff
@@ -28,15 +30,195 @@
+            assert prod_label is not None
+            lhs = KApply(application_label, [contract, KApply(prod_label, arg_vars)])
+            args: List[KInner] = [KEVM.abi_type(input_type, KVariable(input_name)) for input_type, input_name in zip(self.arg_types, self.arg_names)]
+            rhs = KEVM.abi_calldata(self.name, args)
+            opt_conjuncts = [_range_predicate(KVariable(input_name), input_type) for input_name, input_type in zip(self.arg_names, self.arg_types)]
+            conjuncts = [opt_conjunct for opt_conjunct in opt_conjuncts if opt_conjunct is not None]
+            if not conjuncts:
+                ensures = TRUE
+            else:
+                ensures = functools.reduce(lambda x, y: KApply('_andBool_', [x, y]), conjuncts)
+            return KRule(KRewrite(lhs, rhs), ensures=ensures)
+
+    @dataclass
+    class Field:
+        name: str
+        slot: int
+        type: str
+        sort: KSort
+        klabel: KLabel
+        types: FrozenDict
+
+        def __init__(self, name: str, slot: int, type: str, contract_name: str, field_sort: KSort, types: Dict) -> None:
+            self.name = name
+            self.slot = slot
+            self.type = type
+            self.sort = field_sort
+            self.klabel = KLabel(f'field_{contract_name}_{self.name}')
+            self.types = FrozenDict(types)
+
+        def _direct_placement(self, type: str) -> bool:
+            return self.types[type]['encoding'] in {'inplace', 'bytes'} and int(self.types[type]['numberOfBytes']) <= 32
+
+        @property
+        def production(self) -> KProduction:
+            syntax: List[KProductionItem] = [KTerminal(self.name)]
+            curr_type = self.type
+            while True:
+                type = self.types[curr_type]
+                if self._direct_placement(curr_type):
+                    break
+                elif type['encoding'] == 'mapping':
+                    key_type = type['key']
+                    curr_type = type['value']
+                    if self._direct_placement(key_type):
+                        syntax.extend([KTerminal('['), KNonTerminal(Sorts.INT), KTerminal(']')])
+                    else:
+                        raise ValueError(f'Unsupported key type for mapping in field {self.sort}: {key_type}')
+                else:
+                    raise ValueError(f'Unsupported type for encoding in field {self.sort}: {self.type}')
+            return KProduction(self.sort, syntax, klabel=self.klabel)
+
+        def rule(self, contract: KInner, application_label: KLabel) -> KRule:
+            non_terminal_prods = [pitem for pitem in self.production.items if type(pitem) is KNonTerminal]
+            var_names: List[KInner] = [KVariable(f'V{i}') for i, _ in enumerate(non_terminal_prods)]
+            lhs = KApply(application_label, [contract, KApply(self.klabel, var_names)])
+            rhs = KEVM.hashed_location('Solidity', intToken(self.slot), KEVM.intlist(var_names))
```

Comment:
No, this should be the actual correct location. It may be that we need to trim it to the subset of the bytes needed for that storage slot (for example, if it's actually a uint48 it's reading, we may need to do something like `#hashedLocation(...) &Int maxUInt48`. But that would be the remaining adjustment that is made.

`#hashedLocation` is defined in `hashed-locations.md`. Also, see `tests/specs/mcd/storage.k` for examples of using these macros.

Reply:
Correct me if I'm wrong. My understanding of `#hashedLocation` is that it's not a complete implementation of Solidity's storage-layout algorithm but only a helper to deal with dynamic arrays and mappings. Integers, addresses, fixed arrays, and strings require different handling not covered by `#hashedLocation`.

For example, consider the following contract:

~~~Solidity
contract Foo {
    uint256 x;
}
~~~

Here, `x` would have storage location `0` and not `#hashedLocation("Solidity", 0, 0 .IntList)`.
Before the refactoring, the corresponding generated K-rule looked as follows, which seems correct:

~~~
rule  ( Token . x => ( 0 +Int 0 ) )
~~~

But now we get:

~~~
rule  ( Token . x => #hashedLocation ( "Solidity" , 0 , .IntList ) )
~~~

For context, the relevant defining rule of `#hashedLocation` looks as follows:

~~~k
    rule #hashedLocation("Solidity", BASE, OFFSET .IntList) => keccak(#bufStrict(32, OFFSET) ++ #bufStrict(32, BASE))   requires #rangeUInt(256, BASE) andBool #rangeUInt(256, OFFSET)
~~~

---

Moreover, something with `#hashedLocation` doesn't seem right. Generally speaking, I don't think  the rule above captures the correct semantics of storage locations in nested dynamic arrays/mappings. In other words, I don't believe the following equation holds:

~~~
#hashedLocation ( "Solidity" , #hashedLocation ( "Solidity" , BASE , V0 ) , V1 ) ) =
#hashedLocation ( "Solidity" , BASE, V0 V1)
~~~

Reply:
I think the second equality does hold, I think we also have that `#hashedLocation(_, BASE, .IntList) == BASE`, so we have that `0 +Int 0 == #hashedLocation("Solidity", 0, .IntList)`.

I am using the rules defined in `hashed-locations.md` to do this reasoning, specifically https://github.com/runtimeverification/evm-semantics/blob/master/hashed-locations.md

In particular:

```
    rule #hashedLocation(_LANG, BASE, .IntList      ) => BASE
    rule #hashedLocation( LANG, BASE, OFFSET OFFSETS) => #hashedLocation(LANG, #hashedLocation(LANG, BASE, OFFSET .IntList), OFFSETS) requires OFFSETS =/=K .IntList 
```

These are the generic rules that handle 0 offsets or 2+ offsets, and then there are the language specific rules for handling exactly 1 offset:

```
    rule #hashedLocation("Vyper",    BASE, OFFSET .IntList) => keccak(#bufStrict(32, BASE)   ++ #bufStrict(32, OFFSET)) requires #rangeUInt(256, BASE) andBool #rangeUInt(256, OFFSET) 
    rule #hashedLocation("Solidity", BASE, OFFSET .IntList) => keccak(#bufStrict(32, OFFSET) ++ #bufStrict(32, BASE))   requires #rangeUInt(256, BASE) andBool #rangeUInt(256, OFFSET) 
    rule #hashedLocation("Array",    BASE, OFFSET .IntList) => keccak(#bufStrict(32, BASE)) +Word OFFSET                requires #rangeUInt(256, BASE) andBool #rangeUInt(256, OFFSET) 
```